core/state.py

```diff
@@ -1136,6 +1136,53 @@ def _MakeArgvCell(argv):
     return Cell(False, False, False, value.List(items))
 
 
+class ctx_Eval(object):
+    """Push temporary variable frame and override $0, $1, $2, etc."""
+
+    def __init__(self, mem, dollar0, pos_args, vars):
+        # type: (Mem, Optional[str], Optional[List[str]], Optional[Dict[str, value_t]]) -> None
+        self.mem = mem
+        self.pushed_pos_args = False
+        self.pushed_dollar0 = False
+        self.pushed_vars = False
```

Comment:
style nitpick - I think it's clearer to save

```
self.dollar0 = dollar0
self.pos_args = pos_args
self.vars = vars
```

and then in the `__exit__`  you can use the same conditions as in the constructor

Since they are all `is not None` I think it will read clearly